json/src/read.rs

```diff
@@ -609,10 +605,9 @@ fn parse_escape<R: Read>(read: &mut R, scratch: &mut Vec<u8>) -> Result<()> {
 }
 
 fn decode_hex_escape<R: Read>(read: &mut R) -> Result<u16> {
-    let mut i = 0;
     let mut n = 0;
-    while i < 4 && try!(read.peek()).is_some() {
-        n = match try!(read.next()).unwrap_or(b'\x00') {
+    for _ in 0..4 {
+        n = match try!(next_or_eof(read)) {
```

Comment:
This same code dates all the way back to [the very first commit of serde_json](https://github.com/serde-rs/serde/blob/13fd782ac213c8dac4830240f9772c0bdd32f3bb/json.rs#L1366-L1391) in May 2014 and has never been optimized since then, so I'm not sure which performance regression / improvement you are referring to. To my knowledge we have never put optimization effort into escape sequences just because they tend to be uncommon.

The reason the code needs to change in this PR is to return the correct EOF error on an EOF in the middle of a hex escape. The previous code returned an error that was not an EOF error.

In any case, I benchmarked before and after against serde-rs/json-benchmark and there is no difference.